firedrake/slate/slac/compiler.py

```diff
@@ -452,67 +479,86 @@ def facet_integral_loop(cxt_kernel, builder, coordsym, cellfacetsym):
     return loop_stmt, incl
 
 
-def auxiliary_information(builder):
-    """This function generates any auxiliary information regarding special
-    handling of expressions that do not have any integral forms or subkernels
-    associated with it.
+def auxiliary_temporaries(builder, declared_temps):
+    """This function generates auxiliary information regarding special
+    handling of expressions that require creating additional temporaries.
 
     :arg builder: a :class:`KernelBuilder` object that contains all the
                   necessary temporary and expression information.
-
-    Returns: a mapping of the form ``{aux_node: aux_temp}``, where `aux_node`
-             is an already assembled data-object provided as a
-             `firedrake.Function` and `aux_temp` is the corresponding
-             temporary.
-
-             a list of auxiliary statements are returned that contain temporary
+    :arg declared_temps: a `dict` of temporaries that have already been
+                         declared and assigned values. This will be
+                         updated in this method and referenced later
+                         in the compiler.
+    Returns: a list of auxiliary statements are returned that contain temporary
              declarations and any code-blocks needed to evaluate the
              expression.
     """
-    aux_temps = {}
     aux_statements = []
-    for i, exp in enumerate(builder.aux_exprs):
-        if isinstance(exp, Action):
+    for exp in builder.aux_temps:
+        if isinstance(exp, Inverse):
+            if builder._ref_counts[exp] > 1:
+                # Get the temporary for the particular expression
+                result = metaphrase_slate_to_cpp(exp, declared_temps)
+
+                # Now we use the generated result and assign the value to the
+                # corresponding temporary.
+                temp = ast.Symbol("auxT%d" % len(declared_temps))
+                shape = exp.shape
+                aux_statements.append(ast.Decl(eigen_matrixbase_type(shape),
+                                               temp))
+                aux_statements.append(ast.FlatBlock("%s.setZero();\n" % temp))
+                aux_statements.append(ast.Assign(temp, result))
+
+                # Update declared temps
+                declared_temps[exp] = temp
+
+        elif isinstance(exp, Action):
+            # Action computations are relatively inexpensive, so
+            # we don't waste memory space on creating temps for
+            # these expressions. However, we must create a temporary
+            # for the actee coefficient (if we haven't already).
```

Comment:
Yes. Do you think I should create temps for actions? I did previously but decided against it. It's a fairly straightforward thing to add back in.